hkdf.go

```diff
@@ -119,6 +127,18 @@ func ExtractHKDF(h func() hash.Hash, secret, salt []byte) ([]byte, error) {
 		return nil, err
 	}
 
+	// If calling code specifies nil salt, replace it with a buffer of hashLen
+	// zeros, as specified in RFC 5896 and as OpenSSL EVP_KDF-HKDF documentation
+	// instructs. Take a slice of a preallocated buffer to avoid allocating new
+	// buffer per call.
+	if salt == nil {
+		hlen := h().Size()
+		if hlen > len(hkdfAllZerosSalt) {
+			return nil, errors.New("nil salt is not supported: hash output length exceeds internal buffer size")
+		}
+		salt = hkdfAllZerosSalt[:hlen]
```

Comment:
What I had in mind was keeping `hkdfAllZerosSalt` at 64 bytes (to fit the longest that we know can happen, for now) but handle this case gracefully:

```suggestion
		if hlen > len(hkdfAllZerosSalt) {
			salt = make([]byte, hlen)
		} else {
			salt = hkdfAllZerosSalt[:hlen]
		}
```

With that fallback in place, then at worst, if someone adds a new big hash to this repo later and doesn't update the size of `hkdfAllZerosSalt`, performance might be lower for that new hash. If the caller cares, they can even fix it themselves by passing in their own preallocated salt.

(A test that checks `h().Size() <= len(hkdfAllZerosSalt)` for all supported hashes would be nice, but IMO it's not critical for this PR to add it.)